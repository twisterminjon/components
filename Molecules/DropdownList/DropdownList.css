.dropdown-list {
  display: block;
  width: 100%;
  max-width: 100%;
}

.dropdown-list-wrapper {
  display: flex;
  justify-content: space-between;
  align-items: baseline;
}
