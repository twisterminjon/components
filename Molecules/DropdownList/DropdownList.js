import React, { Component } from 'react'
import PropTypes from 'prop-types'

import { Dropdown as SemanticDropdown } from 'semantic-ui-react'

import InputLabel from '../../Atoms/InputLabel/InputLabel'
import InputMessage from '../../Atoms/InputMessage/InputMessage'
import PatientData from '../PatientData/PatientData'
import RequiredLabel from '../../Atoms/RequiredLabel/RequiredLabel'

import './DropdownList.css'

export default class DropdownList extends Component {
  static propTypes = {
    /** name for the input */
    name: PropTypes.string.isRequired,
    /** The current selected value in the dropdown */
    value: PropTypes.string.isRequired,
    /** Label text for the input selected */
    label: PropTypes.string,
    /** Error text if hasError */
    errorMessage: PropTypes.string,
    /** Flag to determine if errorMessage renders */
    hasError: PropTypes.bool,
    /** True displays REQUIRED above input field */
    required: PropTypes.bool,
    /** If true, the input will shown in display only mode */
    displayOnly: PropTypes.bool,
  }
  static defaultProps = {
    label: '',
    errorMessage: '',
    hasError: false,
    displayOnly: false,
  }

  render() {
    const {
      name,
      label,
      errorMessage,
      hasError,
      required,
      style,
      className,
      displayOnly,
      value,
      defaultValue,
      ...rest
    } = this.props
    if (displayOnly) {
      // value is the id, we need to lookup the text in options.
      const option = rest.options.filter(option => option.key === value)
      // sanity check to make sure we got one
      const displayValue = option.length > 0 ? option[0].text : 'Langauge not found'

      return <PatientData title={label} data={displayValue} />
    }

    const requiredRender = required ? <RequiredLabel /> : ''

    const labelId = name + '-label'
    const inputId = name + '-input'
    const hintId = name + '-hint'

    return (
      <div style={style} className={`dropdown-list ${className}`}>
        <div className="dropdown-list-wrapper">
          <InputLabel label={label} htmlFor={inputId} id={labelId} />
          {requiredRender}
        </div>
        <SemanticDropdown
          value={value}
          error={hasError}
          id={inputId}
          aria-labelledby={labelId}
          aria-describedby={hintId}
          {...rest}
        />
        <InputMessage message={errorMessage} show={hasError} />
      </div>
    )
  }
}
