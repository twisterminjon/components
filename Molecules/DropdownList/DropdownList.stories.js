import React from 'react'

import { storiesOf } from '@storybook/react'
import { withInfo } from '@storybook/addon-info'
import centered from '@storybook/addon-centered/react'
import { withKnobs, boolean, text } from '@storybook/addon-knobs'

import DropdownList from './DropdownList'

const stories = storiesOf('Provider/v1/Molecules', module)
stories.addDecorator(withKnobs)

stories.addDecorator(centered).add(
  'DropdownList',
  withInfo()(() => {
    const labelLabel = 'label'
    const labelDefaultValue = 'Dreaming'
    const labelValue = text(labelLabel, labelDefaultValue)

    const placeholderLabel = 'placeholder'
    const placeholderDefaultValue = 'Where do dreams come from'
    const placeholderValue = text(placeholderLabel, placeholderDefaultValue)

    const errorLabel = 'error'
    const errorDefaultValue = 'Oopsy, I did it again'
    const errorValue = text(errorLabel, errorDefaultValue)

    const showErrorLabel = 'showError'
    const showErrorDefaultValue = false
    const showErrorValue = boolean(showErrorLabel, showErrorDefaultValue)

    const requiredLabel = 'required'
    const requiredDefaultValue = false
    const requiredValue = boolean(requiredLabel, requiredDefaultValue)

    const options = [
      { key: '1', value: '?', text: 'The netherworld' },
      { key: '2', value: 'Life', text: 'Life Itself' },
    ]

    return (
      <div style={{ width: 400 }}>
        <DropdownList
          name="dropdown"
          label={labelValue}
          placeholder={placeholderValue}
          errorMessage={errorValue}
          hasError={showErrorValue}
          required={requiredValue}
          options={options}
          fluid
          multiple
          selection
        />
      </div>
    )
  })
)
