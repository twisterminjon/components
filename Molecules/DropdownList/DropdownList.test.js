import React from 'react'
import DropdownList from './DropdownList'

describe('DropdownList', () => {
  const name = 'dropdown'
  const label = 'Klingon'
  const value = 'Worf'
  const fluid = true
  const list = 'list detail'
  const listData = ['Worf', 'Picard', 'Data']
  const placeholder = 'Star Trek'
  const error = 'It is busted'

  it('matches the snapshot', () => {
    const wrapper = window.shallow(
      <DropdownList
        name={name}
        label={label}
        value={value}
        fluid={fluid}
        list={list}
        listData={listData}
        placeholder={placeholder}
        errorMessage={error}
        hasError={false}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('passes the label prop down', () => {
    const wrapper = window.shallow(
      <DropdownList
        name={name}
        label={label}
        value={value}
        fluid={fluid}
        list={list}
        listData={listData}
        placeholder={placeholder}
        errorMessage={error}
        hasError={false}
      />
    )
    expect(wrapper.find('InputLabel').prop('label')).toEqual(label)
  })

  it('passes the placeholder prop down', () => {
    const wrapper = window.shallow(
      <DropdownList
        name={name}
        label={label}
        value={value}
        fluid={fluid}
        list={list}
        listData={listData}
        placeholder={placeholder}
        errorMessage={error}
        hasError={false}
      />
    )
    expect(wrapper.find('Dropdown').prop('placeholder')).toEqual(placeholder)
  })

  it('passes the hasError prop down', () => {
    const wrapper = window.shallow(
      <DropdownList
        name={name}
        label={label}
        value={value}
        fluid={fluid}
        list={list}
        listData={listData}
        placeholder={placeholder}
        errorMessage={error}
        hasError={true}
      />
    )
    expect(wrapper.find('Dropdown').prop('error')).toBeTruthy()
    expect(wrapper.find('InputMessage').prop('show')).toBeTruthy()
  })

  it('passes the error prop down', () => {
    const wrapper = window.shallow(
      <DropdownList
        name={name}
        label={label}
        value={value}
        fluid={fluid}
        list={list}
        listData={listData}
        placeholder={placeholder}
        errorMessage={error}
        hasError={false}
      />
    )
    expect(wrapper.find('InputMessage').prop('message')).toEqual(error)
  })
})
