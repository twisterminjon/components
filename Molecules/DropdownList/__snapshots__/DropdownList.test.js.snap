// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownList matches the snapshot 1`] = `
<div
  className="dropdown-list undefined"
>
  <div
    className="dropdown-list-wrapper"
  >
    <InputLabel
      htmlFor="dropdown-input"
      id="dropdown-label"
      label="Klingon"
    />
  </div>
  <Dropdown
    additionLabel="Add "
    additionPosition="top"
    aria-describedby="dropdown-hint"
    aria-labelledby="dropdown-label"
    closeOnBlur={true}
    closeOnEscape={true}
    deburr={false}
    error={false}
    fluid={true}
    icon="dropdown"
    id="dropdown-input"
    list="list detail"
    listData={
      Array [
        "Worf",
        "Picard",
        "Data",
      ]
    }
    minCharacters={1}
    noResultsMessage="No results found."
    openOnFocus={true}
    placeholder="Star Trek"
    renderLabel={[Function]}
    searchInput="text"
    selectOnBlur={true}
    selectOnNavigation={true}
    value="Worf"
    wrapSelection={true}
  />
  <InputMessage
    message="It is busted"
    show={false}
  />
</div>
`;
